include/gc/Transforms/Utils/ValueUtils.h

```diff
@@ -53,6 +53,27 @@ Value flattenMemref(PatternRewriter &rewriter, Location loc, Value srcMemref);
 // Return true if the memref has shared memory space.
 bool hasSharedMemSpace(mlir::Value memref);
 
+// Go through all parent 'memref.subview' ops for the given `memref`
+// and return the folded offsets of all subviews and the root memref.
+std::tuple<SmallVector<Value>, Value>
+computeSubviewOffsets(PatternRewriter &rewriter, Location loc, Value memref);
+
+// Return the strides of the memref
+SmallVector<OpFoldResult> getMemrefStrides(PatternRewriter &rewriter,
+                                           Location loc, Value memref);
```

Comment:
Maybe pass a vector reference as the function argument to avoid copying the entire vector?

Reply:
I think RVO is guaranteed in our case. We return the result vector straight away from `getMixedValues(...)` or `memref::computeStridesIRBlock(...)`
```c++
SmallVector<OpFoldResult> getMemrefStrides(PatternRewriter &rewriter,
                                           Location loc, Value memref) {
  auto type = dyn_cast<MemRefType>(memref.getType());

  auto stridedLayout = dyn_cast<StridedLayoutAttr>(type.getLayout());
  if (stridedLayout) {
    auto strides = stridedLayout.getStrides();
    return getMixedValues(strides, {}, rewriter);
  }

  auto sizes = getMixedValues(type.getShape(), {}, rewriter);
  auto strides = memref::computeStridesIRBlock(loc, rewriter, sizes);
  return strides;
}
```

Reply:
I'm not sure it's guaranteed, but I agree, in this case it does not make sense. Probably in computeSubviewOffsets() it does.

Reply:
I think it's a bit different case. SmallVector allocates memory on stack, you can't just move the data pointer from one vector to another, you need to copy the internal array and all the objects it contains. Compiler could optimise by inlining the function, but it's not guaranteed.

Reply:
Fixed. Now I pass the output vector to `computeSubviewOffsets()` by reference and fill it with the result